comit/src/btsieve.rs

```diff
@@ -172,3 +189,190 @@ where
         have_seen_block || predates_start_of_swap
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use futures::{future, Stream, TryStreamExt};
+    use genawaiter::GeneratorState;
+    use std::{collections::hash_map::RandomState, iter::FromIterator};
+    use tokio::{
+        sync::Mutex,
+        time::{delay_for, Delay},
+    };
+
+    /// A rough estimate of the IO latency for Infura.
+    const INFURA_LATENCY: Duration = Duration::from_secs(4);
+    /// The approximate mining speed of the Ethereum mainnet.
+    const ETHEREUM_MAINNET_MINING_SPEED: Duration = Duration::from_secs(17);
+    /// We use a poll interval of zero for the test because there is no need to
+    /// artificially slow the processing down.
+    const ZERO_POLL_INTERVAL: Duration = Duration::from_secs(0);
+
+    #[tokio::test]
+    async fn newly_mined_blocks_are_processed_within_a_short_amount_of_time() {
+        let blocks = make_blockchain(100, ETHEREUM_MAINNET_MINING_SPEED);
+        let start_of_swap = blocks[20].timestamp;
+        let connector =
+            FakeConnector::new(blocks, 50, INFURA_LATENCY, ETHEREUM_MAINNET_MINING_SPEED);
+
+        let gen = fetch_blocks_since(&connector, start_of_swap, ZERO_POLL_INTERVAL);
+        let yielded_blocks = fallible_generator_to_try_stream(gen)
+            .map_ok(|b| b.number)
+            .try_take_while(|n| future::ready(Ok(*n != 52)))
+            .try_collect::<Vec<_>>();
+
+        let processed_blocks = tokio::time::timeout(Duration::from_secs(60), yielded_blocks)
+            .await
+            .expect("block 52 to be yielded within 60 seconds")
```

Comment:
Do you want to check it goes backward too?

Reply:
We already check this with our e2e tests and with the integration tests of btsieve. We could in theory replace those tests with what we have here but I don't think it is worth it.